Foundation/URLSession/URLSessionTask.swift

```diff
@@ -682,16 +683,49 @@ extension _ProtocolClient : URLProtocolClient {
             if let protectionSpace = createProtectionSpace(response) {
-                //TODO: Fetch and set proposed credentials if they exist
-                let authenticationChallenge = URLAuthenticationChallenge(protectionSpace: protectionSpace, proposedCredential: nil,
-                                                                     previousFailureCount: task.previousFailureCount, failureResponse: response, error: nil,
-                                                                     sender: urlProtocol as! _HTTPURLProtocol)
-                task.previousFailureCount += 1
-                self.urlProtocol(urlProtocol, didReceive: authenticationChallenge)
+
+                func proceed(proposing credential: URLCredential?) {
+                    let proposedCredential: URLCredential?
+                    let last = task._protocolLock.performLocked { task._lastCredentialUsedFromStorageDuringAuthentication }
+                    
+                    if last?.credential != credential {
+                        proposedCredential = credential
+                    } else {
+                        proposedCredential = nil
+                    }
+                    
+                    let authenticationChallenge = URLAuthenticationChallenge(protectionSpace: protectionSpace, proposedCredential: proposedCredential,
+                                                                             previousFailureCount: task.previousFailureCount, failureResponse: response, error: nil,
+                                                                             sender: urlProtocol as! _HTTPURLProtocol)
+                    task.previousFailureCount += 1
+                    self.urlProtocol(urlProtocol, didReceive: authenticationChallenge)
+                }
+                
+                if let storage = session.configuration.urlCredentialStorage {
+                    storage.getCredentials(for: protectionSpace, task: task) { (credentials) in
+                        if let credentials = credentials,
+                            let firstKeyLexicographically = credentials.keys.sorted().first {
+                            proceed(proposing: credentials[firstKeyLexicographically])
+                        } else {
+                            storage.getDefaultCredential(for: protectionSpace, task: task) { (credential) in
+                                proceed(proposing: credential)
+                            }
+                        }
+                    }
+                } else {
+                    proceed(proposing: nil)
+                }
+                
                 return
             }
         }
         
+        if let storage = session.configuration.urlCredentialStorage,
+           let last = task._protocolLock.performLocked({ task._lastCredentialUsedFromStorageDuringAuthentication }),
+           last.credential.persistence != .none && last.credential.persistence != .forSession {
```

Comment:
`URLCredentialStorage.set(_, for:, task:)` should take care of `.none`. I don't understand completely why `.forSession` is also discarded here. A session is suppose to last until the credential storage stops being used, if I understand correctly.

Reply:
Strategic implementation choice: all storage is best effort, so we can choose not to support `.forSession` storage (because we do not have a safe way to associate a credential with a session right now).

Reply:
I will have to double check the documentation/Darwin implementation, but I always interpreted .forSession as “only in memory”. I also think that this check might be better in the cache, since someone can implement externally  a cache that does .forSession, but not if this check is done here. I think the checks should be in the cache and keep the responsibilities separated.

Reply:
Actually, let me see if I can retrofit it without putting the rest of the patch train too much behind.

Reply:
See comment below — I mistook what `.forSession` meant. We can just allow `forSession` storage.

Reply:
… the same way as we allow `persistent`.